internal/monitors/statsd/statsd.go

```diff
@@ -0,0 +1,161 @@
+package statsd
+
+import (
+	"net"
+	"strconv"
+	"strings"
+	"sync/atomic"
+)
+
+type statsDListener struct {
+	ipAddr         string
+	port           uint16
+	tcp            bool
+	udpConn        *net.UDPConn
+	prefix         string
+	converters     []*converter
+	metricBuffer   []string
+	shutdownCalled int32
+}
+
+type statsDMetric struct {
+	metricName string
+	metricType string
+	value      float64
+	dimensions map[string]string
+}
+
+func (sl *statsDListener) Listen() error {
+	if sl.tcp {
+		return sl.listenTCP()
+	}
+
+	return sl.listenUDP()
+}
+
+func (sl *statsDListener) listenUDP() error {
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{
+		IP:   net.ParseIP(sl.ipAddr),
+		Port: int(sl.port),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	sl.udpConn = conn
+	return nil
+}
+
+func (sl *statsDListener) listenTCP() error {
+	return nil
+}
+
+func (sl *statsDListener) FetchMetrics() []*statsDMetric {
+	rawMetrics := make([]string, len(sl.metricBuffer))
+
+	copy(rawMetrics, sl.metricBuffer)
+	sl.metricBuffer = nil
+
+	parsed := parseMetrics(rawMetrics, sl.converters, sl.prefix)
+
+	return parsed
+}
+
+func (sl *statsDListener) Read() {
+	chData := make(chan []byte)
+
+	if sl.tcp {
+		go sl.readTCP(chData)
+	} else {
+		go sl.readUDP(chData)
+	}
+
+	for data := range chData {
+		sl.metricBuffer = append(sl.metricBuffer, strings.Split(string(data), "\n")...)
+	}
+}
+
+func (sl *statsDListener) readTCP(chData chan []byte) {
+}
+
+func (sl *statsDListener) readUDP(chData chan []byte) {
+	// UDP needs to receive data packet by packet. Max packet size is 65535 for now.
+	buf := make([]byte, 65536)
+	for {
+		// Exit the loop if the connection is closed
+		if atomic.LoadInt32(&sl.shutdownCalled) > 0 {
```

Comment:
Only downside to doing it up here is that it will log an error on monitor shutdown that isn't really an error.  It should do this check in the `if err != nil` block below before logging out an error.

Reply:
Thought wouldn't even get to the `if err != nil` but I see it gets there even after turning on the flag. Guess the flag is still not on when the loop checks for it for the first time after the connection is closed. Moved the flag check into the error block.